### PR TITLE
Add C++ code for Closest Pair problem in Sweep Line module

### PR DESCRIPTION
**Pull request by brebenelmihnea**
_If any of the below don't apply to this Pull Request, mark the checkbox as done._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.

### CODE REVIEW

**Review by dustin-miao — CHANGES_REQUESTED** [review]
Please put the code in the right section under the divide and conquer solution (take a look at [the D&Q modul](https://usaco.guide/plat/DC-SRQ) as an example)

File: content/5_Plat/Sweep_Line.mdx [hunk 1]
```diff
@@ -67,6 +67,90 @@ Extending Solution 1, we can use a set instead of divide and conquer. Once again
...
+pair<Point, Point> strip_solve(vector<Point> &points) {
+	pair<Point, Point> ans = INF;
+	for (int i = 0; i < (int)points.size(); i++)
+		for (int j = i + 1; j < (int)points.size() && j - i < 9; j++)
+			ans = min(ans, {points[i], points[j]});
+	return ans;
+}
+
+pair<Point, Point> solve(vector<Point> &points, int l, int r) {
+	if (r - l + 1 <= 3) {
```
Comment (dustin-miao) [inline content/5_Plat/Sweep_Line.mdx:107]: The divide and conquer solution doesn't require any special handling of small ranges. Try removing this and seeing if it works.

File: content/5_Plat/Sweep_Line.mdx [hunk 2]
```diff
@@ -67,6 +67,90 @@ Extending Solution 1, we can use a set instead of divide and conquer. Once again
...
+};
+
+const pair<Point, Point> INF = {{-1e9, -1e9}, {1e9, 1e9}};
+long double dist(pair<Point, Point> a) {
+	long double d1 = a.first.x - a.second.x;
+	long double d2 = a.first.y - a.second.y;
+	return sqrt(d1 * d1 + d2 * d2);
+}
+
+pair<Point, Point> min(pair<Point, Point> a, pair<Point, Point> b) {
```
Comment (dustin-miao) [inline content/5_Plat/Sweep_Line.mdx:94]: Overloading the `std::min` function like this is confusing. Use a more descriptive name for this function.

File: content/5_Plat/Sweep_Line.mdx [hunk 3]
```diff
@@ -67,6 +67,90 @@ Extending Solution 1, we can use a set instead of divide and conquer. Once again
...
+
+pair<Point, Point> solve(vector<Point> &points, int l, int r) {
+	if (r - l + 1 <= 3) {
+		pair<Point, Point> ans = INF;
+		for (int i = l; i < r; i++)
+			for (int j = i + 1; j < r; j++)
+				ans = min(ans, {points[i], points[j]});
+		return ans;
+	}
+	int mid = (l + r) >> 1;
```
Comment (dustin-miao) [inline content/5_Plat/Sweep_Line.mdx:114]:
```suggestion
	int mid = (l + r) / 2;
```

File: content/5_Plat/Sweep_Line.mdx (hunk 1 above)
Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx]: You're right. It  removed that part.

File: content/5_Plat/Sweep_Line.mdx (hunk 2 above)
Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx]: I changed it to `get_closest_points` .

File: content/5_Plat/Sweep_Line.mdx (hunk 3 above)
Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx]: I changed it to `/ 2`

**Comment by brebenelmihnea** [comment]
> Please put the code in the right section under the divide and conquer solution (take a look at [the D&Q module](https://usaco.guide/plat/DC-SRQ) as an example).

I moved it under `Solution 1`.

**Review by dustin-miao — APPROVED** [review]

File: content/5_Plat/Sweep_Line.mdx [hunk 4]
```diff
@@ -55,6 +55,84 @@ To achieve the desired $\mathcal{O}(n)$ complexity per layer, we need to be able
...
+
+struct Point {
+	long double x, y;
+	bool operator<(const Point &other) {
+		if (x == other.x) return y < other.y;
+		return x < other.x;
+	}
+};
+
+const pair<Point, Point> INF = {{-1e9, -1e9}, {1e9, 1e9}};
```
Comment (SansPapyrus683) [inline content/5_Plat/Sweep_Line.mdx:74]:
```suggestion
const pair<Point, Point> INF{{-1e9, -1e9}, {1e9, 1e9}};
```
does this work

File: content/5_Plat/Sweep_Line.mdx [hunk 5]
```diff
@@ -55,6 +55,84 @@ To achieve the desired $\mathcal{O}(n)$ complexity per layer, we need to be able
...
+pair<Point, Point> strip_solve(vector<Point> &points) {
+	pair<Point, Point> ans = INF;
+	for (int i = 0; i < (int)points.size(); i++)
+		for (int j = i + 1; j < (int)points.size() && j - i < 9; j++)
+			ans = get_closest_points(ans, {points[i], points[j]});
+	return ans;
+}
+
+pair<Point, Point> solve(vector<Point> &points, int l, int r) {
+	if (l == r) return INF;
```
Comment (SansPapyrus683) [inline content/5_Plat/Sweep_Line.mdx:95]: braces

File: content/5_Plat/Sweep_Line.mdx [hunk 6]
```diff
@@ -55,6 +55,84 @@ To achieve the desired $\mathcal{O}(n)$ complexity per layer, we need to be able
...
+	long double d2 = a.first.y - a.second.y;
+	return sqrt(d1 * d1 + d2 * d2);
+}
+
+pair<Point, Point> get_closest_points(pair<Point, Point> a,
+                                      pair<Point, Point> b) {
+	return (dist(a) < dist(b) ? a : b);
+}
+
+pair<Point, Point> strip_solve(vector<Point> &points) {
```
Comment (SansPapyrus683) [inline content/5_Plat/Sweep_Line.mdx:86]: doc comments for all these functions

File: content/5_Plat/Sweep_Line.mdx [hunk 7]
```diff
@@ -55,6 +55,84 @@ To achieve the desired $\mathcal{O}(n)$ complexity per layer, we need to be able
...
+			scanf("%Lf %Lf", &x, &y);
+			v.push_back({x, y});
+		}
+		sort(v.begin(), v.end());
+		pair<Point, Point> ans = solve(v, 0, v.size());
+		printf("%0.2Lf %0.2Lf %0.2Lf %0.2Lf\n", ans.first.x, ans.first.y,
+		       ans.second.x, ans.second.y);
+	}
+
+	return 0;
```
Comment (SansPapyrus683) [inline content/5_Plat/Sweep_Line.mdx:129]:
```suggestion
```

File: content/5_Plat/Sweep_Line.mdx [hunk 8]
```diff
@@ -55,6 +55,84 @@ To achieve the desired $\mathcal{O}(n)$ complexity per layer, we need to be able
...
+
+pair<Point, Point> strip_solve(vector<Point> &points) {
+	pair<Point, Point> ans = INF;
+	for (int i = 0; i < (int)points.size(); i++)
+		for (int j = i + 1; j < (int)points.size() && j - i < 9; j++)
+			ans = get_closest_points(ans, {points[i], points[j]});
+	return ans;
+}
+
+pair<Point, Point> solve(vector<Point> &points, int l, int r) {
```
Comment (SansPapyrus683) [inline content/5_Plat/Sweep_Line.mdx:94]: ngl i feel like you could've just put this function in `main`

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/5_Plat/Sweep_Line.mdx (hunk 4 above)
Comment (dustin-miao) [reply content/5_Plat/Sweep_Line.mdx:74]: Why is this any better?

File: content/5_Plat/Sweep_Line.mdx (hunk 8 above)
Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx:94]: As a lambda function ? I  would be too much for the stack memory.

File: content/5_Plat/Sweep_Line.mdx (hunk 4 above)
Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx]: Done.

File: content/5_Plat/Sweep_Line.mdx (hunk 5 above)
Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx]: Done

File: content/5_Plat/Sweep_Line.mdx (hunk 7 above)
Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx:129]: Do you want to remove `return 0` ?

File: content/5_Plat/Sweep_Line.mdx (hunk 6 above)
Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx]: Done

File: content/5_Plat/Sweep_Line.mdx (hunk 7 above)
Comment (SansPapyrus683) [reply content/5_Plat/Sweep_Line.mdx:133]: yeah

File: content/5_Plat/Sweep_Line.mdx (hunk 8 above)
Comment (SansPapyrus683) [reply content/5_Plat/Sweep_Line.mdx]: no like move the code directly into it without all the parameters
unless it recurses? if so I didn’t see it

Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx]: Yeah, it recursive.

File: content/5_Plat/Sweep_Line.mdx (hunk 7 above)
Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx]: I removed it

File: content/5_Plat/Sweep_Line.mdx [hunk 9]
```diff
@@ -55,6 +55,86 @@ To achieve the desired $\mathcal{O}(n)$ complexity per layer, we need to be able
...
+	return sqrt(d1 * d1 + d2 * d2);
+}
+
+pair<Point, Point> get_closest_points(pair<Point, Point> a,
+                                      pair<Point, Point> b) {
+	return (dist(a) < dist(b) ? a : b);
+}
+
+pair<Point, Point> strip_solve(
+    vector<Point> &points) {  // Brute force for points with near mid_point
```
Comment (SansPapyrus683) [inline content/5_Plat/Sweep_Line.mdx:87]:
```suggestion
/** Brute force for points with near mid_point */
pair<Point, Point> strip_solve(vector<Point> &points) {
```

File: content/5_Plat/Sweep_Line.mdx [hunk 10]
```diff
@@ -55,6 +55,86 @@ To achieve the desired $\mathcal{O}(n)$ complexity per layer, we need to be able
...
+	pair<Point, Point> ans = INF;
+	for (int i = 0; i < (int)points.size(); i++)
+		for (int j = i + 1; j < (int)points.size() && j - i < 9; j++)
+			ans = get_closest_points(ans, {points[i], points[j]});
+	return ans;
+}
+
+pair<Point, Point> solve(vector<Point> &points, int l,
+                         int r) {  // Solve the problem for range [l, r]
```
Comment (SansPapyrus683) [inline content/5_Plat/Sweep_Line.mdx:96]:
```suggestion
/** Solve the problem for range [l, r] */
pair<Point, Point> solve(vector<Point> &points, int l, int r) {
```

File: content/5_Plat/Sweep_Line.mdx [hunk 11]
```diff
@@ -55,6 +55,86 @@ To achieve the desired $\mathcal{O}(n)$ complexity per layer, we need to be able
...
+                                      pair<Point, Point> b) {
+	return (dist(a) < dist(b) ? a : b);
+}
+
+pair<Point, Point> strip_solve(
+    vector<Point> &points) {  // Brute force for points with near mid_point
+	pair<Point, Point> ans = INF;
+	for (int i = 0; i < (int)points.size(); i++)
+		for (int j = i + 1; j < (int)points.size() && j - i < 9; j++)
+			ans = get_closest_points(ans, {points[i], points[j]});
```
Comment (SansPapyrus683) [inline content/5_Plat/Sweep_Line.mdx:91]: braces

File: content/5_Plat/Sweep_Line.mdx [hunk 12]
```diff
@@ -55,6 +55,86 @@ To achieve the desired $\mathcal{O}(n)$ complexity per layer, we need to be able
...
+
+const pair<Point, Point> INF{{-1e9, -1e9}, {1e9, 1e9}};
+long double dist(pair<Point, Point> a) {
+	long double d1 = a.first.x - a.second.x;
+	long double d2 = a.first.y - a.second.y;
+	return sqrt(d1 * d1 + d2 * d2);
+}
+
+pair<Point, Point> get_closest_points(pair<Point, Point> a,
+                                      pair<Point, Point> b) {
```
Comment (SansPapyrus683) [inline content/5_Plat/Sweep_Line.mdx:82]: use const references here

File: content/5_Plat/Sweep_Line.mdx [hunk 13]
```diff
@@ -55,6 +55,86 @@ To achieve the desired $\mathcal{O}(n)$ complexity per layer, we need to be able
...
+struct Point {
+	long double x, y;
+	bool operator<(const Point &other) {
+		if (x == other.x) return y < other.y;
+		return x < other.x;
+	}
+};
+
+const pair<Point, Point> INF{{-1e9, -1e9}, {1e9, 1e9}};
+long double dist(pair<Point, Point> a) {
```
Comment (SansPapyrus683) [inline content/5_Plat/Sweep_Line.mdx:75]: and here

File: content/5_Plat/Sweep_Line.mdx [hunk 14]
```diff
@@ -55,6 +55,86 @@ To achieve the desired $\mathcal{O}(n)$ complexity per layer, we need to be able
...
+
+struct Point {
+	long double x, y;
+	bool operator<(const Point &other) {
+		if (x == other.x) return y < other.y;
+		return x < other.x;
+	}
+};
+
+const pair<Point, Point> INF{{-1e9, -1e9}, {1e9, 1e9}};
```
Comment (SansPapyrus683) [inline content/5_Plat/Sweep_Line.mdx:74]:
```suggestion
const pair<Point, Point> INF{{-1e9, -1e9}, {1e9, 1e9}};

```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/5_Plat/Sweep_Line.mdx (hunk 9 above)
Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx]: Done. I think the `pre-commit.ci `bot keeps changing this.

File: content/5_Plat/Sweep_Line.mdx (hunk 10 above)
Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx]: Done

File: content/5_Plat/Sweep_Line.mdx (hunk 11 above)
Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx:92]: Done.

File: content/5_Plat/Sweep_Line.mdx (hunk 12 above)
Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx]: Done

File: content/5_Plat/Sweep_Line.mdx (hunk 13 above)
Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx]: Done.

File: content/5_Plat/Sweep_Line.mdx (hunk 14 above)
Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx:74]: Done.

File: content/5_Plat/Sweep_Line.mdx [hunk 15]
```diff
@@ -55,6 +55,89 @@ To achieve the desired $\mathcal{O}(n)$ complexity per layer, we need to be able
...
+	long double d2 = a.first.y - a.second.y;
+	return sqrt(d1 * d1 + d2 * d2);
+}
+
+pair<Point, Point> get_closest_points(const pair<Point, Point> &a,
+                                      const pair<Point, Point> &b) {
+	return (dist(a) < dist(b) ? a : b);
+}
+
+/* Brute force for points with near mid_point */
```
Comment (SansPapyrus683) [inline content/5_Plat/Sweep_Line.mdx:87]:
```suggestion
/** Brute force for points with near mid_point */
```
what's `mid_point`? clarify that as well

Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx]: Done. I said - `the median point in the sorted array`

File: content/5_Plat/Sweep_Line.mdx [hunk 16]
```diff
@@ -55,6 +55,92 @@ To achieve the desired $\mathcal{O}(n)$ complexity per layer, we need to be able
...
+	pair<Point, Point> ans_right = solve(points, mid + 1, r);
+	pair<Point, Point> ans;
+
+	ans = get_closest_points(ans_left, ans_right);
+	long double delta = dist(ans);
+
+	Point mid_point = points[mid];
+	vector<Point> strip;
+	for (int i = l; i < r; i++)
+		if (abs(points[i].x - mid_point.x) <= delta) strip.push_back(points[i]);
```
Comment (SansPapyrus683) [inline content/5_Plat/Sweep_Line.mdx:117]: bro i thought i asked for braces 💀

Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx]: I put again braces. I think the pre-commit bot changes the file.

Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx]: Do you see the braces now ?

File: content/5_Plat/Sweep_Line.mdx [hunk 17]
```diff
@@ -55,6 +55,94 @@ To achieve the desired $\mathcal{O}(n)$ complexity per layer, we need to be able
...
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+struct Point {
+	long double x, y;
+	bool operator<(const Point &other) {
+		if (x == other.x) return y < other.y;
```
Comment (SansPapyrus683) [inline content/5_Plat/Sweep_Line.mdx:68]: still no braces here

File: content/5_Plat/Sweep_Line.mdx [hunk 18]
```diff
@@ -55,6 +55,94 @@ To achieve the desired $\mathcal{O}(n)$ complexity per layer, we need to be able
...
+	// The smallest distance in range [mid+1, r]
+	pair<Point, Point> ans_right = solve(points, mid + 1, r);
+	pair<Point, Point> ans;
+
+	ans = get_closest_points(ans_left, ans_right);
+	long double delta = dist(ans);
+
+	Point mid_point = points[mid];
+	vector<Point> strip;
+	for (int i = l; i < r; i++)
```
Comment (SansPapyrus683) [inline content/5_Plat/Sweep_Line.mdx:116]: or here
when i ask for braces i want braces _everywhere_

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: content/5_Plat/Sweep_Line.mdx (hunk 17 above)
Comment (brebenelmihnea) [reply content/5_Plat/Sweep_Line.mdx]: I added braces.

**Pull request merged**